sematic/future_context.py

```diff
@@ -0,0 +1,88 @@
+# Standard Library
+from contextlib import contextmanager
+from dataclasses import dataclass
+from importlib import import_module
+from typing import Any, Optional, Type
+
+
+@dataclass(frozen=True)
+class SematicContext:
+    """Contextual information about the execution of the current Sematic function
+    Attributes
+    ----------
+    id:
+        The id of the future for the current execution. For cloud executions, this
+        is equivalent to the id for the existing run.
+    root_id:
+        The id of the root future for a resolution. For cloud executions, this is
+        equivalent to the id for the root run.
+    resolver_class_path:
+        The import path for the resolver being used.
+    """
+
+    id: str
+    root_id: str
+    resolver_class_path: str
+
+    def resolver_class(self) -> Type[Any]:
+        module_name, resolver_name = self.resolver_class_path.rsplit(".", maxsplit=1)
+        module = import_module(module_name)
+        resolver_class = getattr(module, resolver_name, None)
+        if resolver_class is None:
+            raise ImportError(f"No class named '{resolver_name}' in {module_name}")
+        return resolver_class
+
+
+_current_context: Optional[SematicContext] = None
+
+
+@contextmanager
+def set_context(ctx: SematicContext):
+    """Context manager to execute a Sematic function with a global context enabled.
+
+    This should only be called by Sematic itself, not by end-users.
+
+    Parameters
+    ----------
+    ctx:
+        The context that the function should be executed with
+    """
+    global _current_context
+
+    # _current_context might not be None if a Sematic resolution is started
+    # from within a Sematic func. This is not a pattern we encourage, but
+    # shouldn't fail.
```

Comment:
Or should it?

Reply:
Generally I'm for being lenient in what we accept, and for being flexible in allowing users to do "advanced" things in order to get around limitations.

But I'm concerned about starting resolutions inside resolutions, as the imbricated interactions with the Server and DB might get complicated.

A resolution started inside a run will not be bound to the enclosing Future func in any way, as it will be kicked off while the Future is being CREATED by executing Python code. The implications would be numerous.

I would be against supporting this use case.

Reply:
K, will make it an error! Prior to this PR, that would have worked. But now we are making it fail. So *technically* that makes this an API change. But only to discourage something we didn't want anyway so it's probably ok.